sstables/compress.cc

```diff
@@ -33,7 +33,7 @@ enum class mask_type : uint8_t {
 
 // size_bits cannot be >= 64
 static inline uint64_t make_mask(uint8_t size_bits, uint8_t offset, mask_type t) noexcept {
-    const uint64_t mask = ((1 << size_bits) - 1) << offset;
+    const uint64_t mask = ((uint64_t(1) << size_bits) - 1) << offset;
```

Comment:
Note if it does happen, the server is dead, since it means we allocate 4GB. A shard usually only has 7GB.

Reply:
I don't know what this code is responsible for and I wasn't able to understand it with a cursory look. But somehow this triggered UBSAN in the CI for v1 of this patch (https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2811/), and the node apparently didn't die.

Reply:
Clarification: UBSAN was triggered by a new test, which used huge compression chunks. Not by the `zstd.cc` patch itself.

Reply:
A test wouldn't kill the node, but a production node that suddenly tried to allocate 4GB would die a horrible death.

Reply:
I think you are confusing this piece of code with something else. AFAIK the numbers here aren't related to any allocation sizes.

As far as I understand, this code is used for bitpacking offsets of compression chunks in the data file. In particular, in the case which triggers UBSAN `size_bits` is a bit width of a file offset. So >32 is perfectly possible here. Nothing allocation-related.

Reply:
I don't understand exactly what's being encoded here, but I know that the voodoo magic in https://github.com/scylladb/scylladb/blob/b9d88a36011b88788df82e2d2d5036a8e6fec298/sstables/segmented_compress_params.hh says that all you need to trigger this UB is to set the compression chunk size to >=2MiB.

Reply:
The 2MiB bound follows from this line: https://github.com/scylladb/scylladb/blob/b9d88a36011b88788df82e2d2d5036a8e6fec298/sstables/segmented_compress_params.hh#L49

Reply:
Ah, you're right, that's part of CompressionInfo.db management.

Reply:
@denesb please verify, you wrote the code. There's no question the fix is correct, just if we need to backport it. Likely not, since we'd have seen a bug here long ago.

Reply:
I wrote this a lifetime ago. The size passed to this method is always `_segmented_offset_size_bits`, calculated here:
```
https://github.com/scylladb/scylladb/blob/b9d88a36011b88788df82e2d2d5036a8e6fec298/sstables/compress.cc#L171
```
After plugging in various chunk sizes I found that we can indeed get a value of 33 with the largest supported chunk size of 2**30 (1GiB). But the one below that (512MiB) is already below this.

Reply:
> the size passed to this method is always _segmented_offset_size_bits

@denesb That's not true. Sometimes the size passed to this method is `_segment_base_offset_size_bits`, which is calculated using the table at https://github.com/scylladb/scylladb/blob/b9d88a36011b88788df82e2d2d5036a8e6fec298/sstables/segmented_compress_params.hh#L31.

> that we can indeed get a value of 33 with the largest supported chunk size of 2**30 (1GiB). But the one below that (512MiB) is already below this.

The table I linked dictates a relationship between chunk size and `_segment_base_offset_size_bits`. You can see that for a chunk size with bit width >=21, (meaning chunk size >=  2MiB), `_segment_base_offset_size_bits` is >=32, triggering the UB. So the problem starts with 2MiB, not 512MiB.

You can easily verify that's the case by setting the compression chunk size to 2MiB in some test and running it in debug mode. It will trip UBSAN.

Reply:
(I did run such test.)

Reply:
Right, I missed the usage in `write()`. Well, then there is not much to discuss, we have to backport this. I wonder how no one hit this yet.

Reply:
>  I wonder how no one hit this yet.

2MiB chunks are pretty big. Maybe nobody uses them?

I also don't see any tests which use 2 MiB chunks. I think the biggest chunk size we test is 1 MiB, in SCT: https://github.com/scylladb/scylla-cluster-tests/blob/bfb8ebf684134257de920f4e7d97eb834716d91a/sdcm/nemesis.py#L2260

Reply:
But do note that with chunks <2MiB the result will be well-defined, but it still may be wrong, because information is lost in the second shift (`<< offset`).

E.g. with `size_bits == 25` and `offset == 7`, `mask` is going to be equal to `0xffffffff'ffffff80`, instead of the intended `0x0000007f'ffffff80`, due to sign extension.

Reply:
Wrong result from this method would cause failure to read compressed sstables. So I think it is unlikely we've hit this in production. I think it is best to write an exploratory unit test mapping up the bounds of chunk sizes, within which the function is correct and incorrect.

Reply:
> So I think it is unlikely we've hit this in production.

So do I.

> I think it is best to write an exploratory unit test mapping up the bounds of chunk sizes, within which the function is correct and incorrect.

I don't understand this proposal.

Reply:
> I don't understand this proposal.

I'd like to have a unit test which uses compression with carious chunk sizes. In particular I want this test to answer the question of what is the smallest chunk size that already causes problems before the fix. But we should also test the smallest and largest chunk sizes to make sure we don't regress in the future either.
Maybe you already have a test which accomplishes all this, in which case there is no additional changes needed.

Reply:
2MB chunks will not work well.

Reply:
@denesb I think I would need to spend a considerable amount of effort to write such a test. It would require gaining understanding of the CompressionInfo code.

In particular, the test which originally uncovered this UB was still passing accidentally despite the bug, due to the exact argument values — the write and read masks were wrong, but because they are _both_ equally wrong, and the written value (0) happened to fit in these masks, the result came out OK. So the test would require ensuring that actually multiple chunks/buckets/something are written, or something like that, so that the loss of information due to bad masks is actually meaningful.

Reply:
(And this may require something more involved than just running any workload with chunk sizes configured so-and-so).

Reply:
The thing is, I'm not sure what values go to these masks.

E.g. if the value that goes there is always 0, there's a dreadful possibility that the bug actually happens in production and writes some garbage to compressioninfo, but the bugs in the read path and write path happen to cancel each other, and fixing the bug will cause a regression because the corrupted data will no longer be readable...

Reply:
You can find the design for the compression of compressioninfo here: https://github.com/scylladb/scylladb/issues/1946 (https://github.com/scylladb/scylladb/issues/1946#issuecomment-305999782).
The entire machinery is used to store in a compressed form the compression-info component of already used sstables, it is used exclusively on reads. So no chance of a mistake in `make_mask()` causing corrupt sstables. Also, you do not have to be afraid that a fix to this code will cause problems with old nodes. The in-memory state doesn't persist between restarts.

A wrong value calculated by `make_mask()` *could* go unnoticed if it resulted in looking up the wrong chunk on the disk. But this is highly unlikely, as that would almost surely would result in a parse failure, or even an earlier decompress failure.